File: sync-bridge-nestjs/src/health/health.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Customer } from '../sync/entities/customer.entity';

@Injectable()
export class HealthService {
  constructor(
    @InjectRepository(Customer)
    private readonly customerRepo: Repository<Customer>,
  ) {}

  async healthCheck() {
    const result = { read: false, write: false, timestamp: new Date().toISOString() };

    try {
      await this.customerRepo.findOne({ where: {} });
      result.read = true;
    } catch (e) {
      result.read = false;
    }

    try {
      const temp = this.customerRepo.create({
        email: 'healthcheck@example.com',
        firstName: 'Health',
        lastName: 'Check',
        defaultCurrency: 'USD',
      });
      const saved = await this.customerRepo.save(temp);
      await this.customerRepo.remove(saved);
      result.write = true;
    } catch (e) {
      result.write = false;
    }

    return result;
  }
}

File: sync-bridge-csharp/SyncBridgeCsharp/Services/HealthService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using SyncBridgeCsharp.Data;
using SyncBridgeCsharp.Models;

namespace SyncBridgeCsharp.Services;

public interface IHealthService
{
    Task<Dictionary<string, object>> HealthCheckAsync();
}

public class HealthService : IHealthService
{
    private readonly SyncDbContext _context;

    public HealthService(SyncDbContext context)
    {
        _context = context;
    }

    public async Task<Dictionary<string, object>> HealthCheckAsync()
    {
        bool readOk = false;
        bool writeOk = false;

        try
        {
            // Read check
            await _context.Customers.AnyAsync();
            readOk = true;
        }
        catch
        {
            // Ignored
        }

        try
        {
            // Write check: insert and immediately delete a temporary customer
            var temp = new Customer
            {
                Email = $"healthcheck_{Guid.NewGuid()}@example.com", // use unique to avoid collision if run concurrently
                FirstName = "Health",
                LastName = "Check",
                DefaultCurrency = "USD"
            };

            _context.Customers.Add(temp);
            await _context.SaveChangesAsync();

            _context.Customers.Remove(temp);
            await _context.SaveChangesAsync();

            writeOk = true;
        }
        catch
        {
            // Ignored
        }

        return new Dictionary<string, object>
        {
            { "read", readOk },
            { "write", writeOk },
            { "timestamp", DateTimeOffset.UtcNow.ToString("o") }
        };
    }
}
